heroku_connect/models.py

```diff
@@ -0,0 +1,342 @@
+from django.conf import settings
+from django.db import models
+from django.db.models import F, OuterRef, Subquery
+from django.db.models.functions import Coalesce
+
+from heroku_connect.db import models as hc_models
+
+
+class TriggerLogQuerySet(models.QuerySet):
+    """A QuerySet for trigger log models."""
+
+    def failed(self):
+        """Filter for log records with sync failures."""
+        return self.filter(state=TriggerLogState.FAILED)
+
+    def related_to(self, instance):
+        """Filter for all log objects of the same connected model as the given instance."""
+        return self.filter(table_name=instance.table_name, record_id=instance.record_id)
+
+
+class TriggerLogAction:
+    """Type of change that a trigger log object represents."""
+
+    INSERT = 'INSERT'
+    """A new connected model instance was created locally."""
+    UPDATE = 'UPDATE'
+    """A connected model instance was updated locally."""
+    DELETE = 'DELETE'
+    """A connected model instance was deleted locally."""
+
+    @classmethod
+    def choices(cls):
+        return tuple((getattr(cls, name), name) for name in dir(cls) if name.isupper())
```

Comment:
Don't you think this is utterly complex for:
```
('INSERT', 'INSERT'), ('UPDATE', 'UPDATE'), ('DELETE','DELETE')
```
?

Reply:
If you want to use the class as django choices, I would rather recommend to build a helper for that.
```python
def cls_to_choices(cls):
    return cls.__dict__.items()
```

Reply:
Ideally I'd like Django to have a choices-enum class! ;P My idea was to:

- have the choices adapt automatically when the class attributes change;
- not duplicate the much bigger list of states on the other class;
- make it clear on the class itself that the attributes will be used as choices;
- avoid using a dependency  for it.

So I'd prefer to leaving the method on the class instead of using a helper from somewhere else. I could introduce a super class to avoid duplicating the method code. Using `__dict__` is fine, but it still needs to at least filter out names starting with `_` or maybe only `__`.

Reply:
I'll just ause a `dict` instead.